rafthttp/peer.go

```diff
@@ -15,158 +15,169 @@
 package rafthttp
 
 import (
-	"errors"
+	"log"
 	"net/http"
-	"sync"
 	"time"
 
+	"github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context"
 	"github.com/coreos/etcd/etcdserver/stats"
 	"github.com/coreos/etcd/pkg/types"
 	"github.com/coreos/etcd/raft/raftpb"
 )
 
 const (
-	appRespBatchMs = 50
-	propBatchMs    = 10
-
+	DialTimeout      = time.Second
 	ConnReadTimeout  = 5 * time.Second
 	ConnWriteTimeout = 5 * time.Second
+
+	recvBufSize = 4096
 )
 
 type peer struct {
-	sync.Mutex
-
-	id  types.ID
-	cid types.ID
-
-	tr http.RoundTripper
-	// the url this sender post to
-	u  string
-	r  Raft
-	fs *stats.FollowerStats
-
-	batcher     *Batcher
-	propBatcher *ProposalBatcher
+	id types.ID
+
+	msgAppWriter *streamWriter
+	writer       *streamWriter
+	pipeline     *pipeline
+
+	sendc   chan raftpb.Message
+	recvc   chan raftpb.Message
+	newURLc chan string
+	pausec  chan struct{}
+	resumec chan struct{}
+	stopc   chan struct{}
+	done    chan struct{}
+}
 
-	pipeline *pipeline
-	stream   *stream
+func startPeer(tr http.RoundTripper, u string, local, to, cid types.ID, r Raft, fs *stats.FollowerStats, errorc chan error) *peer {
+	p := &peer{
+		id:           to,
+		msgAppWriter: startStreamWriter(fs),
+		writer:       startStreamWriter(fs),
+		pipeline:     newPipeline(tr, u, to, cid, fs, errorc),
+		sendc:        make(chan raftpb.Message),
+		recvc:        make(chan raftpb.Message, recvBufSize),
+		newURLc:      make(chan string),
+		pausec:       make(chan struct{}),
+		resumec:      make(chan struct{}),
+		stopc:        make(chan struct{}),
+		done:         make(chan struct{}),
+	}
+	go func() {
+		msgAppReader := startStreamReader(tr, u, streamTypeMsgApp, local, to, cid, p.recvc)
+		reader := startStreamReader(tr, u, streamTypeGeneral, local, to, cid, p.recvc)
+		var paused bool
+
+		// non-blocking run loop
+		for {
+			select {
+			case m := <-p.sendc:
+				if paused {
+					continue
+				}
+				writec, name, size := p.pick(m)
+				select {
+				case writec <- m:
+				default:
+					log.Printf("peer: dropping %s to %s since %s with %d-size buffer is blocked",
+						m.Type, p.id, name, size)
+				}
+			case mm := <-p.recvc:
+				if mm.Type == raftpb.MsgApp {
```

Comment:
yep. that is the original design. To do it, i can make writer close the underlying connection if higher-term msgapp comes.